File: ios/Autolock/Bluelink/BluelinkModels.swift
import Foundation

/// Persisted Bluelink session: tokens plus the vehicle identifiers needed to
/// address commands at a specific car. Cached across app relaunches (the
/// background App Intent runs in a fresh process each time) so a normal lock
/// trigger costs a token refresh + one command call rather than a full
/// login + vehicle lookup every time -- both to be fast and to conserve
/// Bluelink's daily rate limit.
struct BluelinkSession: Codable, Equatable {
    var accessToken: String
    var refreshToken: String
    var tokenExpiry: Date
    var vehicle: BluelinkVehicleInfo?

    /// Matches bluelinky's ~10s safety margin before treating a token as
    /// needing refresh.
    func isAccessTokenValid(now: Date = Date()) -> Bool {
        now.addingTimeInterval(10) < tokenExpiry
    }
}

struct BluelinkVehicleInfo: Codable, Equatable {
    let regId: String
    let vin: String
    let generation: Int
    let nickname: String
}

enum BluelinkLockAction: String {
    case lock
    case unlock

    /// Path segment under the API base for this action.
    var endpointPath: String {
        switch self {
        case .lock: return "rcs/rdo/off"
        case .unlock: return "rcs/rdo/on"
        }
    }
}

enum BluelinkOrderStatus {
    case pending
    case success
    case failed
    case unknown
}

// MARK: - Wire response shapes

struct BluelinkTokenResponse: Decodable {
    let accessToken: String
    let refreshToken: String
    let expiresIn: Double
    let errorCode: String?
    let errorMessage: String?

    enum CodingKeys: String, CodingKey {
        case accessToken = "access_token"
        case refreshToken = "refresh_token"
        case expiresIn = "expires_in"
        case errorCode
        case errorMessage
    }

    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        accessToken = try container.decodeIfPresent(String.self, forKey: .accessToken) ?? ""
        refreshToken = try container.decodeIfPresent(String.self, forKey: .refreshToken) ?? ""
        errorCode = try container.decodeIfPresent(String.self, forKey: .errorCode)
        errorMessage = try container.decodeIfPresent(String.self, forKey: .errorMessage)

        // expires_in has been observed as both a JSON number and a numeric
        // string across API responses; accept either.
        if let asDouble = try? container.decode(Double.self, forKey: .expiresIn) {
            expiresIn = asDouble
        } else if let asString = try? container.decode(String.self, forKey: .expiresIn),
                  let parsed = Double(asString) {
            expiresIn = parsed
        } else {
            expiresIn = 0
        }
    }
}

struct BluelinkEnrolledVehiclesResponse: Decodable {
    struct Entry: Decodable {
        struct Details: Decodable {
            let regid: String
            let nickName: String
            let vin: String
            let vehicleGeneration: String?
            let enrollmentStatus: String?
        }
        let vehicleDetails: Details
    }
    let enrolledVehicleDetails: [Entry]?
    let errorCode: String?
    let errorMessage: String?
}

struct BluelinkActionStatusResponse: Decodable {
    let status: String?
    let errorCode: String?
    let errorMessage: String?
}

File: android/app/src/main/java/com/sendprobe/autolock/bluelink/BluelinkModels.kt
package com.sendprobe.autolock.bluelink

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.doubleOrNull

/**
 * Persisted Bluelink session: tokens plus the vehicle identifiers needed to
 * address commands at a specific car. Cached across app relaunches (a
 * notification trigger fires in a fresh process) so a normal lock trigger
 * costs a token refresh + one command call rather than a full login +
 * vehicle lookup every time -- both to be fast and to conserve Bluelink's
 * daily rate limit.
 */
@Serializable
data class BluelinkSession(
    val accessToken: String,
    val refreshToken: String,
    val tokenExpiryMillis: Long,
    val vehicle: BluelinkVehicleInfo? = null
) {
    /** Matches bluelinky's ~10s safety margin before treating a token as needing refresh. */
    fun isAccessTokenValid(nowMillis: Long = System.currentTimeMillis()): Boolean =
        nowMillis + 10_000L < tokenExpiryMillis
}

@Serializable
data class BluelinkVehicleInfo(
    val regId: String,
    val vin: String,
    val generation: Int,
    val nickname: String
)

enum class BluelinkLockAction(val endpointPath: String) {
    LOCK("rcs/rdo/off"),
    UNLOCK("rcs/rdo/on")
}

enum class BluelinkOrderStatus {
    PENDING, SUCCESS, FAILED, UNKNOWN
}

// MARK: - Wire request shapes

@Serializable
internal data class LoginRequestBody(
    val username: String,
    val password: String
)

@Serializable
internal data class RefreshRequestBody(
    val username: String,
    val password: String,
    @SerialName("grant_type") val grantType: String = "refresh_token",
    @SerialName("refresh_token") val refreshToken: String
)

@Serializable
internal data class LockActionRequestBody(
    @SerialName("userName") val userName: String,
    val vin: String
)

// MARK: - Wire response shapes

@Serializable
internal data class BluelinkTokenResponse(
    @SerialName("access_token") val accessToken: String? = null,
    @SerialName("refresh_token") val refreshToken: String? = null,
    // Observed as both a JSON number and a numeric string across API
    // responses; kept as a raw element and coerced by expiresInSecondsOrZero.
    @SerialName("expires_in") val expiresIn: JsonElement? = null,
    val errorCode: String? = null,
    val errorMessage: String? = null
) {
    val expiresInSecondsOrZero: Double
        get() = (expiresIn as? JsonPrimitive)?.doubleOrNull ?: 0.0
}

@Serializable
internal data class EnrolledVehiclesResponse(
    val enrolledVehicleDetails: List<Entry>? = null,
    val errorCode: String? = null,
    val errorMessage: String? = null
) {
    @Serializable
    internal data class Entry(val vehicleDetails: Details)

    @Serializable
    internal data class Details(
        val regid: String,
        val nickName: String,
        val vin: String,
        val vehicleGeneration: String? = null,
        val enrollmentStatus: String? = null
    )
}

@Serializable
internal data class BluelinkActionStatusResponse(
    val status: String? = null,
    val errorCode: String? = null,
    val errorMessage: String? = null
)
